[参考資料](http://magic.iswbm.com/zh/latest/)

## 1."import this"だけじゃない

In [1]:
import antigravity

In [45]:
import __hello__

Hello world!


In [46]:
import this

The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!


## 2. ...

In [3]:
...

Ellipsis

In [7]:
print(type(...))
print(id(...)) # ...はellipsisクラスのsingleton
print(id(...))

<class 'ellipsis'>
4352958848
4352958848


In [8]:
bool(...)

True

### [使い道](https://stackoverflow.com/questions/772124/what-does-the-ellipsis-object-do)

In [10]:
# passとして使う
def foo():
    pass

def bar():
    ...

foo()
bar()

In [ ]:
# typingとして使う
from typing import Callable

def partial(func: Callable[..., str], *args) -> Callable[..., str]:
    # Body

[PEP484](https://www.python.org/dev/peps/pep-0484/#callable)

In [11]:
# Numpyでslice操作
import numpy as np

foo = np.arange(16).reshape(2, 2, 4)

foo

array([[[ 0,  1,  2,  3],
        [ 4,  5,  6,  7]],

       [[ 8,  9, 10, 11],
        [12, 13, 14, 15]]])

In [14]:
foo[:, :, 2]

array([[ 2,  6],
       [10, 14]])

In [12]:
foo[..., 2] # 同じ結果

array([[ 2,  6],
       [10, 14]])

## 3. shellのendを真似したいなら...(使わないでください)

In [15]:
__builtins__.end = None


def my_abs(x):
    if x > 0:
        return x
    else:
        return -x
    end
end

print(my_abs(10))
print(my_abs(-10))

10
10


## 4. promptを自分で定義する

In [ ]:
```python
>>> import sys
>>> sys.ps1
'>>> '
>>> sys.ps2
'... '
>>> sys.ps3
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'sys' has no attribute 'ps3'
>>> sys.ps1 = "海賊王に俺はなる!>>>"
海賊王に俺はなる!>>>sys.ps2 = "肉~"
海賊王に俺はなる!>>>for i in range(2):
肉~     print("サンジ、腹減った")
肉~
サンジ、腹減った
サンジ、腹減った
海賊王に俺はなる!>>>
```

## 5.and & or

In [18]:
(2 or 3) * (5 and 6 and 7)

14

In [21]:
"" and "b" # andは左から右へスキャンして、最初のFalseを戻る


''

In [22]:
"a" and "b" # 全てTrueなら、最後の値を戻る

'b'

In [24]:
"" or "a" # orは左から右へスキャンして、最初のTrueを戻る

'a'

In [26]:
"" or [] or None or {} # 全てFalseなら、最後の値を戻る

{}

## 6. リストを結合する

In [67]:
a = [1, 2]
b = [3, 4]
c = [5, 6]
sum((a, b, c), [])

[1, 2, 3, 4, 5, 6]

In [28]:
[*a, *b, *c]

[1, 2, 3, 4, 5, 6]

In [29]:
from itertools import chain

list(chain(a, b, c))

[1, 2, 3, 4, 5, 6]

## 7. \[-5, 256]

In [30]:
a = -6
b = -6

c = -5
d = -5

print(a is b)
print(c is d)

False
True


> CPythonでは、\[-5, 256]間の整数は事前に定義されたオブジェクト  
> 他の数字と違う

## 8. 文字列のintern

In [33]:
s1 = "hello" # internと言うのは同じ文字列はメモリにただ一つ保存され
s2 = "hello"
s1 is s2

True

In [34]:
s3 = 'hell o' # スペースある場合は、internは起動しない
s4 = 'hell o'
s3 is s4

False

## 9. dict()と{}

In [40]:
timeit -n 1000000 -r 5 dict()

112 ns ± 4.2 ns per loop (mean ± std. dev. of 5 runs, 1000000 loops each)


In [41]:
timeit -n 1000000 -r 5 {}

21.6 ns ± 1.42 ns per loop (mean ± std. dev. of 5 runs, 1000000 loops each)


`{}`は`dict()`より早い

In [42]:
from dis import dis
dis("{}")

  1           0 BUILD_MAP                0
              2 RETURN_VALUE


In [44]:
dis('dict()')

  1           0 LOAD_NAME                0 (dict)
              2 CALL_FUNCTION            0
              4 RETURN_VALUE


## 10."e"と"е"

In [50]:
valuе = 12
print(value)

NameError: name 'value' is not defined

> 原因は上の"valuе"の"е"は普通の"e"ではない

In [52]:
ord("e")

101

In [53]:
ord("е")

1077

> 使い道？  
> いたずら！

## 11. "site"

In [55]:
import sys

sys.path

['/Users/maozhongfu/GitHub/Python',
 '/Users/maozhongfu/.vscode/extensions/ms-python.python-2020.10.332292344/pythonFiles',
 '/Users/maozhongfu/.vscode/extensions/ms-python.python-2020.10.332292344/pythonFiles/lib/python',
 '/Library/Frameworks/Python.framework/Versions/3.8/lib/python38.zip',
 '/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8',
 '/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/lib-dynload',
 '',
 '/Users/maozhongfu/Library/Python/3.8/lib/python/site-packages',
 '/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages',
 '/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/aeosa',
 '/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/IPython/extensions',
 '/Users/maozhongfu/.ipython',
 '/Users/maozhongfu/.vscode/extensions/ms-python.python-2020.10.332292344/pythonFiles/vscode_datascience_helpers/dataframes']

In [56]:
!python3 -m site

sys.path = [
    '/Users/maozhongfu/GitHub/Python',
    '/Users/maozhongfu/.vscode/extensions/ms-python.python-2020.10.332292344/pythonFiles',
    '/Users/maozhongfu/.vscode/extensions/ms-python.python-2020.10.332292344/pythonFiles/lib/python',
    '/Library/Frameworks/Python.framework/Versions/3.8/lib/python38.zip',
    '/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8',
    '/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/lib-dynload',
    '/Users/maozhongfu/Library/Python/3.8/lib/python/site-packages',
    '/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages',
    '/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/aeosa',
]
USER_BASE: '/Users/maozhongfu/Library/Python/3.8' (exists)
USER_SITE: '/Users/maozhongfu/Library/Python/3.8/lib/python/site-packages' (exists)
ENABLE_USER_SITE: True


## 12. http server

```shell
cd <path> # シェアしたいフォルダまで移動
ipconfig # 自分のipを把握他人に教える
python3 -m http.server <port number> # port numberでサーバーを立ち上げる
# その後 "http://<ip adress>:<port number>"でアクセルできる
```

## 13. pydoc

```python
python3 -m pydoc -p <port number>
```

## 14. ":=" (python3.8以上)

### if..else..

In [73]:
age = 22
if age > 20:
    print("お酒を飲めてもいい")

お酒を飲めてもいい


In [74]:
if (age:=22) > 20:
    print("お酒を飲めてもいい")

お酒を飲めてもいい


### while

In [ ]:
file = open("demo.txt", "r")
while True:
    line = file.readline()
    if not line:
        break
    print(line.strip())

In [ ]:
file = open("demo.txt", "r")
while (line := file.readline()):
    print(line.strip())

### comprehension

In [ ]:
scores = [22,54,75,89]
valid_scores = [
   longFunction(n)
   for n in scores
   if longFunction(n) # longFunctionは2回実行される
]

In [ ]:
scores = [22,54,75,89]
valid_scores = [
   result for n in scores
   result := longFunction(n) # 一回だけ
]

## 15. decorator

### パラメータありの関数のdecorator

In [3]:
def say_hello(country:str = None):
    def wrapper(func):
        def deco(*args, **kwargs):
            if country == "China":
                print("你好!")
            elif country == "Japan":
                print("こんにちは！")
            else:
                print("Winter is Coming!")
            
            func(*args, **kwargs)
        
        return deco
    return wrapper

@say_hello("China")
def 诸葛孔明():
    ...

@say_hello("Japan")
def 桃太郎():
    pass

@say_hello()
def JonSnow():
    pass

In [4]:
诸葛孔明()
桃太郎()
JonSnow()

你好!
こんにちは！
Winter is Coming!


### パラメータ無しのクラスdecorator

In [6]:
class logger(object):
    def __init__(self, func):
        self._func = func
    
    def __call__(self, *args, **kwargs):
        print(f"[INFO]: function {self._func.__name__} is running")
        return self._func(*args, **kwargs)

@logger
def say(something):
    print(f"say {something}")

say("Hello")

[INFO]: function say is running
say Hello


### パラメータありのクラスdecorator

In [9]:
class logger(object):
    def __init__(self, level="INFO"):
        self._level = level
    
    def __call__(self, func):
        def wrapper(*args, **kwargs):
            print(f"[{self._level}]: function {func.__name__} is running")
            func(*args, **kwargs)
        
        return wrapper

@logger(level="DEBUG")
def say(something):
    print(f"say {something}")

say("Hello") 

[DEBUG]: function say is running
say Hello


### クラスのdecorator

In [11]:
def singleton(cls):
    _instance = {}

    def inner():
        if cls not in _instance:
            _instance[cls] = cls()
        return _instance[cls]
    return inner

@singleton
class Test(object):
    def __init__(self):
        pass

foo = Test()
bar = Test()
print(id(foo) == id(bar))

True


## 16.[lru_cache](https://docs.python.org/ja/3/library/functools.html#functools.lru_cache)

In [15]:
def fib(n):
    if n < 2:
        return n
    return fib(n - 2) + fib(n - 1)

In [16]:
from functools import lru_cache

@lru_cache()
def fib_with_cache(n):
    if n < 2:
        return n
    return fib_with_cache(n - 2) + fib_with_cache(n - 1)

In [19]:
time fib(20)

CPU times: user 2.65 ms, sys: 35 µs, total: 2.69 ms
Wall time: 2.75 ms


6765

In [22]:
time fib_with_cache(2000)

CPU times: user 1.35 ms, sys: 8 µs, total: 1.35 ms
Wall time: 1.38 ms


4224696333392304878706725602341482782579852840250681098010280137314308584370130707224123599639141511088446087538909603607640194711643596029271983312598737326253555802606991585915229492453904998722256795316982874482472992263901833716778060607011615497886719879858311468870876264597369086722884023654422295243347964480139515349562972087652656069529806499841977448720155612802665404554171717881930324025204312082516817125

## 17.print to file

In [24]:
with open("test.txt", mode="w") as f:
    print("hello world", file=f)

In [25]:
!cat test.txt

hello world


In [26]:
!rm test.txt

## 18. for..else.., try..else..

In [28]:
def check_item(source:list, target:str):
    for item in source:
        if item == target:
            print("あった！")
            break
    else:
        print("無かった")

source=["Apple", "Huawei", "Xiaomi"]

In [29]:
check_item(source, "Huawei")

あった！


In [30]:
check_item(source, "Sharp")

無かった


In [31]:
def test_try_else(attr1 = None):
    try:
        if attr1:
            pass
        else:
            raise
    except:
        print("Exception occurred...")
    else:
        print("No Exception occurred...")

In [32]:
test_try_else()

Exception occurred...


In [33]:
test_try_else("~")

No Exception occurred...


## 19.numba

In [35]:
!pip install numba

     |████████████████████████████████| 2.2 MB 3.8 MB/s 
     |████████████████████████████████| 18.4 MB 19.5 MB/s 


In [43]:
import numba as nb

@nb.jit()
def nb_sum(a):
    Sum = 0
    for i in range(len(a)):
        Sum += a[i]
    return Sum

def py_sum(a):
    Sum = 0
    for i in range(len(a)):
        Sum += a[i]
    return Sum

In [44]:
import numpy as np
a = np.linspace(0,100,100)

%timeit np.sum(a) # numpy
%timeit sum(a) # pythonのsum関数
%timeit nb_sum(a) # numba
%timeit py_sum(a) # 上のやつ

4.1 µs ± 106 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
17.8 µs ± 193 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
372 ns ± 0.67 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
23.8 µs ± 475 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


## 20. 里程標日付

* Python 1.0 1994
* Python 2.Xの最期 2020/1/1
* Python 3.0 2008/12/3

## 21.PEP

* [PEP 0](https://www.python.org/dev/peps/): 全てのPEPのIndex
* [PEP 1](https://www.python.org/dev/peps/pep-0001/): PEPの目標とガイドライン
* [PEP 8](https://www.python.org/dev/peps/pep-0008/): スタイルガイド
* [PEP 20](https://www.python.org/dev/peps/pep-0020/): Pythonの禅
* [PEP 404](https://www.python.org/dev/peps/pep-0404/): 永遠に来ないPython 2.8
* [PEP 484](https://www.python.org/dev/peps/pep-0484/): 型ヒント
* [PEP 498](https://www.python.org/dev/peps/pep-0498/): f-string

## 22.CPythonだけじゃない

* [Stackless Python](https://stackless.readthedocs.io/en/3.6-slp/stackless-python.html)
* [Jython](https://www.jython.org) -> もう死んだ同然
* [IronPython](https://ironpython.net)
* [PyPy](https://www.pypy.org)
* [MicroPython](https://micropython.org)
* [Cython](https://cython.org)